cachix/src/Cachix/Client/Commands.hs

```diff
@@ -256,18 +260,41 @@ retryText :: RetryStatus -> Text
 retryText retrystatus =
   if rsIterNumber retrystatus == 0
     then ""
-    else "(retry #" <> show (rsIterNumber retrystatus) <> ") "
+    else color Yellow $ "retry #" <> show (rsIterNumber retrystatus) <> " "
 
 pushStrategy :: Store -> Maybe Token -> PushOptions -> Text -> BinaryCache.CompressionMethod -> StorePath -> PushStrategy IO ()
 pushStrategy store authToken opts name compressionMethod storePath =
   PushStrategy
     { onAlreadyPresent = pass,
       on401 = handleCacheResponse name authToken,
       onError = throwM,
-      onAttempt = \retrystatus size -> do
-        path <- decodeUtf8With lenientDecode <$> storePathToPath store storePath
-        -- we append newline instead of putStrLn due to https://github.com/haskell/text/issues/242
-        putStr $ retryText retrystatus <> "compressing using " <> T.toLower (show compressionMethod) <> " and pushing " <> path <> " (" <> humanSize (fromIntegral size) <> ")\n",
+      onAttempt = \_ _ -> return (),
+      onStream = \retrystatus size -> do
+        path <- liftIO $ decodeUtf8With lenientDecode <$> storePathToPath store storePath
+        let hSize = toS $ humanSize $ fromIntegral size
+            bar = color Blue "[:bar] " <> toS (retryText retrystatus) <> toS path <> " (:percent of " <> hSize <> ")"
+            barLength = T.length $ T.replace ":percent" "  0%" (T.replace "[:bar]" "" (toS bar))
+
+        progressBar <-
+          liftIO $
+            newProgressBar
+              def
+                { pgTotal = fromIntegral size,
+                  -- https://github.com/yamadapc/haskell-ascii-progress/issues/24
+                  pgWidth = 20 + barLength,
+                  pgOnCompletion = Just $ color Green "✓ " <> toS path <> " (" <> hSize <> ")",
+                  pgFormat = bar
+                }
+
+        isTerminal <- liftIO $ hIsTerminalDevice stdout
+        unless isTerminal $ do
```

Comment:
I would expect an `if isTerminal` here; two branches, instead of doing the progress bar even when there's no terminal.

Reply:
Yeah, me too.